_posts/2018-02-12-Reusable-features-in-React-Redux.md

```diff
@@ -0,0 +1,188 @@
+changed the way I think, it introduced me to the concepts of Functional and Reactive programming 
+and showed me how easy and maintainable the development of a large scale Single Page Applications can be.
+Nightmare of state management became delightful experience and after I started handling my side effects
+with [redux-observabe](https://redux-observable.js.org/) it has really become something special.
+
+Today I'd like to talk about how to create features in React+Redux application in standard
+and prescribed way. What do I mean by standard and prescribed ? Applying the same patters/rules
+over and over so that every part of the codebase looks the same regardless whether
+it was written by *Developer1* or *Developer2*.
+
+For the purpose of this article, I've implemented all the patterns that we will be talking about in this [repository](https://github.com/char0n/react-reusable-features). 
+Feel free to clone it and run it by **npm run start**. I will start by explaining how I approach creation
+of **standard react features** and later we'll get into topic of creating features that are able to create
+instances which can be mounted on different locations in the state. I call these features **reusable features**.
+
+## Standard features
+
+You can imagine a standard feature as a directory that lives in your codebase. 
+I will try to demonstrate what standard feature is on Todo List usecase. 
+The standard feature directory has prescribed structure and usually looks like this:
+
+```
+src/
+ todo-feature/
+   components/ 
+     TodoList.js     
+     TodoListItem.js
+   containers/
+     TodoList.js  
+     TodoListItem.js
+   action.js
+   reducers.js
+   selectors.js
+   utils.js
+   middleware.js
+```   
+
+#### components/
+
+The directory where your components live.
+
+#### containers/
+
+The directory where your containers live. If you don't know what is the difference between the components 
+and containers you should read [this article](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0)
+from Dan Abramov why this separation exists and why it is necessary.
+
+#### actions.js
+
+The file where your [action types](https://redux.js.org/docs/basics/Actions.html#actions) and
+[action creators](https://redux.js.org/docs/basics/Actions.html#action-creators) are defined.
+
+
+#### reducers.js
+
+The file where your [reducers](https://redux.js.org/docs/basics/Reducers.html) are defined. 
+This is also a place where you compose the reducers of your possible sub-features using [combineReducers](https://redux.js.org/docs/api/combineReducers.html).
+
+
+#### selectors.js
+
+The file where you define your selectors. Selectors are pure functions that compose and know how
+to retrieve data from your state. If you're now familiar with the concept I highly recommend looking 
+at [reselect](https://github.com/reactjs/reselect) library. I strongly recommend dividing the selectors into two categories. 
```

Comment:
It could just be `reselect`, but if you add `reselect library` then it will need an article such as `the`.

```diff
- at [reselect](https://github.com/reactjs/reselect) library
+ at the [reselect](https://github.com/reactjs/reselect) library
```